idaes/apps/grid_integration/multiperiod/design_and_operation_models.py

```diff
@@ -0,0 +1,122 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+
+from pyomo.environ import (
+    Var,
+    Param,
+    Binary,
+    Expression,
+    NonNegativeReals,
+    Constraint,
+)
+from functools import reduce
+from idaes.core.base.process_base import declare_process_block_class
+from idaes.core.base.process_base import ProcessBlockData
+from pyomo.common.config import ConfigValue, In
+
+
+@declare_process_block_class("DesignModel")
+class DesignModelData(ProcessBlockData):
+    """
+    Class for containing design model ...
+    """
+
+    CONFIG = ProcessBlockData.CONFIG()
+    CONFIG.declare(
+        "model_func",
+        ConfigValue(
+            doc="Function that builds the design model",
+        ),
+    )
+    CONFIG.declare(
+        "model_args",
+        ConfigValue(
+            default={},
+            doc="Dictionary containing arguments needed for model_func",
+        ),
+    )
+
+    # noinspection PyAttributeOutsideInit
+    def build(self):
+        super().build()
+
+        self.config.model_func(self, **self.config.model_args)
+
+
+@declare_process_block_class("OperationModel")
+class OperationModelData(ProcessBlockData):
+    """
+    Class for containing design model ...
+    """
+
+    CONFIG = ProcessBlockData.CONFIG()
+    CONFIG.declare(
+        "model_func",
+        ConfigValue(
+            doc="Function that builds the design model",
+        ),
+    )
+    CONFIG.declare(
+        "model_args",
+        ConfigValue(
+            default={},
+            doc="Dictionary containing arguments needed for model_func",
+        ),
+    )
+
+    CONFIG.declare(
+        "declare_op_vars",
+        ConfigValue(
+            default=True,
+            domain=In([True, False]),
+            doc="Should op_mode, startup, shutdown vars be defined?",
+        ),
+    )
+    CONFIG.declare(
+        "append_lmp_data",
+        ConfigValue(
+            default=True,
+            domain=In([True, False]),
+            doc="Should LMP data automatically be appended to the model?",
+        ),
+    )
+
+    # noinspection PyAttributeOutsideInit
+    def build(self):
```

Comment:
@andrewlee94 I was just thinking about this, and on the one hand, using derived classes would be more in line with how the rest of IDAES does things. On the other hand, using this `model_func` approach gives the potential of applying either approach: I can formulate quick and dirty models on the fly by using the config option to pass in a build function, or I can go and create a child class. So in a sense, allowing for the current approach would give more flexibility, while eliminating it would be more restrictive on the user (which might be the way to go in most cases, but not necessarily here). 

In any case, I think that any refactor of this formulation should be part of a subsequent PR. I think it is more important for this overall capability to get merged in first, since there has already been published work on this Pricetaker approach (for one).

Reply:
> In any case, I think that any refactor of this formulation should be part of a subsequent PR. I think it is more important for this overall capability to get merged in first, since there has already been published work on this Pricetaker approach (for one).

This approach is backwards. If we know a refactor is coming, it's better to do it before the PR is merged. Otherwise we need to deprecate the old class, leave it in a few cycles, and create a similarly-named new class in order to maintain backwards compatibility.

Reply:
In fact, we don't _know_ a refactor is coming and it is something that is really up for discussion at this point--which is why I said "if any refactor." This PriceTaker class has been delayed for quite some time now, and I don't think we should wait until the discussion settles.